packages/frontend/core/src/mixpanel/events/type.ts

```diff
@@ -0,0 +1,101 @@
+interface QuickSearchOptionEvents {
+  changeLanguage: string;
+}
+
+interface SettingsModuleEvents {
+  quickSearchOption: QuickSearchOptionEvents;
+}
+
+interface CMDKSegmentEvents {
+  settings: SettingsModuleEvents;
+}
+
+interface GeneralPageEvents {
+  cmdk: CMDKSegmentEvents;
+}
+
+interface DocPageEvents {}
+
+interface EdgelessPageEvents {}
+
+interface MixpanelEvents {
+  general: GeneralPageEvents;
+  doc: DocPageEvents;
+  edgeless: EdgelessPageEvents;
+}
+
+type MakeEvent = {
+  [Page in keyof MixpanelEvents]: {
+    [Segment in keyof MixpanelEvents[Page]]: {
+      [Module in keyof MixpanelEvents[Page][Segment]]: {
+        [Control in keyof MixpanelEvents[Page][Segment][Module]]: {
+          [Event in keyof MixpanelEvents[Page][Segment][Module][Control]]: (
+            argv: MixpanelEvents[Page][Segment][Module][Control][Event]
+          ) => void;
+        };
+      };
+    };
+  };
+};
```

Comment:
```suggestion
type KeysOfUnion<T> = T extends T ? keyof T : never;
type ValuesOfUnion<T> = T extends T ? T[keyof T] : never;

type MakeEventLevel<
  Level extends string = '>',
  Type = MixpanelEvents,
> = Level extends '>>>>>>'
  ? {
      [Key in KeysOfUnion<Type>]: (
        argv: Extract<Type, Record<Key, any>>[Key]
      ) => void;
    }
  : {
      [Key in KeysOfUnion<Type> | '$']: MakeEventLevel<
        `${Level}>`,
        Key extends '$'
          ? ValuesOfUnion<Type>
          : Extract<Type, Record<Key, any>>[Key]
      >;
    };

type MakeEvent = MakeEventLevel;
```

Reply:
We can then replace any unknown part with $
```ts
track.$.cmdk.$.quickSearchOption.changeLanguage()
```

Reply:
![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/5c0ymolP9B7QStCsS1RP/09dd0fb7-1a8d-4932-9fb8-e0ee53c777e8.png)

the typing is gonna be in a mess

Reply:
Works fine on my side